qcodes/dataset/dond/do_nd.py

```diff
@@ -34,104 +35,438 @@
 )
 from qcodes.parameters import ParameterBase
+from qcodes.utils import deprecate
 
-from .sweeps import AbstractSweep
+from .sweeps import AbstractSweep, TogetherSweep
 
 LOG = logging.getLogger(__name__)
 
+SweepVarType = Any
+
 
 class ParameterGroup(TypedDict):
     params: tuple[ParamMeasT, ...]
     meas_name: str
-    measured_params: list[res_type]
+
+
+@dataclass
+class ParameterSetEvent:
+    parameter: ParameterBase
+    new_value: float
+    should_set: bool
+    delay: float
+    actions: ActionsT
 
 
 class _Sweeper:
     def __init__(
         self,
-        sweeps: Sequence[AbstractSweep],
+        sweeps: Sequence[AbstractSweep | TogetherSweep],
         additional_setpoints: Sequence[ParameterBase],
     ):
-        self._sweeps = sweeps
         self._additional_setpoints = additional_setpoints
-        self._nested_setpoints = self._make_nested_setpoints()
+        self._sweeps = sweeps
+        self._setpoints = self._make_setpoints_tuples()
+        self._setpoints_dict = self._make_setpoints_dict()
+        self._shapes = self._make_shape(sweeps, additional_setpoints)
 
-        self._post_delays = tuple(sweep.delay for sweep in sweeps)
-        self._params_set = tuple(sweep.param for sweep in sweeps)
-        self._post_actions = tuple(sweep.post_actions for sweep in sweeps)
+    @property
+    def setpoint_dicts(self) -> dict[str, list[Any]]:
+        return self._setpoints_dict
 
-    def _make_nested_setpoints(self) -> np.ndarray:
-        """Create the cartesian product of all the setpoint values."""
-        if len(self._sweeps) == 0:
-            return np.array([[]])  # 0d sweep (do0d)
-        setpoint_values = [sweep.get_setpoints() for sweep in self._sweeps]
-        return self._flatten_setpoint_values(setpoint_values)
+    def _make_setpoints_tuples(
+        self,
+    ) -> tuple[tuple[tuple[SweepVarType, ...] | SweepVarType]]:
+        sweeps = tuple(sweep.get_setpoints() for sweep in self._sweeps)
+        return cast(
+            Tuple[Tuple[Union[Tuple[SweepVarType, ...], SweepVarType]]],
+            tuple(itertools.product(*sweeps)),
+        )
 
-    @staticmethod
-    def _flatten_setpoint_values(setpoint_values: Sequence[np.ndarray]) -> np.ndarray:
-        setpoint_grids = np.meshgrid(*setpoint_values, indexing="ij")
-        flat_setpoint_grids = [np.ravel(grid, order="C") for grid in setpoint_grids]
-        return np.vstack(flat_setpoint_grids).T
+    def _make_single_point_setpoints_dict(self, index: int) -> dict[str, SweepVarType]:
+
+        setpoint_dict = {}
+        values = self._setpoints[index]
+        for sweep, subvalues in zip(self._sweeps, values):
+            if isinstance(sweep, TogetherSweep):
+                for individual_sweep, value in zip(sweep.sweeps, subvalues):
+                    setpoint_dict[individual_sweep.param.full_name] = value
+            else:
+                setpoint_dict[sweep.param.full_name] = subvalues
+        return setpoint_dict
```

Comment:
Imho that would be sad. We don't want to couple the logic of generating a single setpoint to the whole thing

Reply:
> We don't want to couple the logic of generating a single setpoint to the whole thing

why not? the logic is the same, right? and the "whole thing" is generated anyways, so why not take singe points from the whole thing since it exists already?

Reply:
They are now but there is no reason that they always need to be and making them more coupled for no good reason is going the wrong way

Reply:
> there is no reason that they always need to be

oh, that was not clear to me. do i guess it correctly that, for example, those logics won't be the same if the structure of setpoints is more complicated than some nesting of sweeps, but the need for a function that returns "the next dict of setpoint values" would still be needed?

Reply:
I mean as its written now we iterate over the sweeper which uses this method to generate the next events. This is currently implemented using getitem but in the long run it would be more correct as a generator that can only be consumed once. However, all of that is out of scope for this pr but there is no need to make it harder down the line

Reply:
thanks for clarifying, agree!